File: 1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.ts
function arrayRankTransform(arr: number[]): number[] {
    // 배열을 복사한 후 오름차순 정렬
    const sortedArr: number[] = [...arr].sort((a, b) => a - b);

    const rankMap: Map<number, number> = new Map();

    for (const num of sortedArr) {
        if (!rankMap.has(num)) {
            rankMap.set(num, rankMap.size + 1);
        }
    }

    // 원본 배열의 각 값을 순위로 변환
    const result: number[] = new Array(arr.length);

    for (let i = 0; i < arr.length; i++) {
        result[i] = rankMap.get(arr[i])!;
    }

    return result;
}

File: leetcode/1331-rank-transform-of-an-array/1331-rank-transform-of-an-array.java
class Solution {
    public int[] arrayRankTransform(int[] arr) {
        // Clone and sort the array to get unique sorted elements
        int[] sortedArr = arr.clone();
        Arrays.sort(sortedArr);
        
        Map<Integer, Integer> rankMap = new HashMap<>();
        for (int num : sortedArr) {
            rankMap.putIfAbsent(num, rankMap.size() + 1);
        }
        
        // Transform the original array
        int[] result = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            result[i] = rankMap.get(arr[i]);
        }
        
        return result;
    }
}
